obspy/io/shapefile/core.py

```diff
@@ -0,0 +1,282 @@
+# -*- coding: utf-8 -*-
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from future.builtins import *  # NOQA
+from future.utils import native_str
+
+from obspy import Catalog
+from obspy.core.event import Origin, Magnitude
+from obspy.core.inventory import Inventory
+
+try:
+    import gdal
+    from osgeo import ogr, osr
+except ImportError as e:
+    has_GDAL = False
+    IMPORTERROR_MSG = str(e) + (
+        ". ObsPy's write support for shapefiles requires the 'gdal' module "
+        "to be installed in addition to the general ObsPy dependencies.")
+else:
+    has_GDAL = True
+    gdal.UseExceptions()
+
+
+def _write_shapefile(obj, filename, **kwargs):
+    """
+    Write :class:`~obspy.core.event.Catalog` object to a ESRI shapefile.
+
+    :type obj: :class:`~obspy.core.event.Catalog` or
+        :class:`~obspy.core.inventory.Inventory`
+    :param obj: ObsPy object for shapefile output
+    :type filename: str or file
+    :param filename: Filename to write to. According to ESRI shapefile
+        definition, multiple files with the following suffixes will be written:
+        ".shp", ".shx", ".dbj", ".prj".
+    """
+    if not has_GDAL:
+        raise ImportError(IMPORTERROR_MSG)
+    if not filename.endswith(".shp"):
+        filename += ".shp"
+
+    driver = ogr.GetDriverByName(native_str("ESRI Shapefile"))
+
+    driver.DeleteDataSource(filename)
+    data_source = driver.CreateDataSource(filename)
```

Comment:
I based the code on this example: pcjericks.github.io/py-gdalogr-cookbook/layers.html#create-a-new-shapefile-and-add-data

I think I read somewhere that the `Destroy()` calls are flushing the files, can't find it right now though. Maybe it could make sense to put the `Destroy()` calls in a try/finally but that's just guessing..